File: app/src/main/java/ch/bailu/aat/services/cache/elevation/HillshadeColorTable.java
package ch.bailu.aat.services.cache.elevation;

import ch.bailu.aat.services.cache.ObjectHandle;
import ch.bailu.aat.util.AppBroadcaster;
import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.services.background.BackgroundTask;
import ch.bailu.aat.services.dem.tile.MultiCell;

public final class HillshadeColorTable extends ObjectHandle {

    public final static String ID=HillshadeColorTable.class.getSimpleName();

    private final static int MAX_DARKNES=50;
    private final static int TABLE_DIM=500;
    private final static int TABLE_HDIM=TABLE_DIM/2;
    private final static int TABLE_SIZE=TABLE_DIM*TABLE_DIM;

    private final static int MIN_DELTA=-250;
    private final static int MAX_DELTA=240;

    private static final int COLOR=50;
    private static final int GRAY=(COLOR << 16) | (COLOR << 8) | COLOR;

    private final byte[][] table=new byte[TABLE_DIM][TABLE_DIM];

    public HillshadeColorTable() {
        super(ID);
    }


    private boolean isInitialized=false;

    @Override
    public void onInsert(ServiceContext sc) {
        sc.getBackgroundService().process(new TableInitializer());
    }

    @Override
    public void onDownloaded(String id, String url, ServiceContext sc) {

    }

    @Override
    public void onChanged(String id, ServiceContext sc) {

    }

    @Override
    public boolean isReadyAndLoaded() {
        return isInitialized;
    }


    public long getSize() {
        return TABLE_SIZE;
    }


    private static float indexToDelta(int i) {
        return (i-TABLE_HDIM) / 100f;
    }

    private static int deltaToIndex(int d) {
        return (d+TABLE_HDIM);
    }


    private static int cutDelta(int d) {
        d=Math.max(MIN_DELTA, d);
        d=Math.min(MAX_DELTA, d);
        return d;
    }


    public int getColor(final MultiCell mcell) {
        final int x=deltaToIndex(cutDelta(mcell.delta_zx()));
        final int y=deltaToIndex(cutDelta(mcell.delta_zy()));
        final int alpha=table[x][y] & 0xFF;

        return (alpha << 24) | GRAY;
    }


    private final class TableInitializer extends BackgroundTask {
        /**
         * Source:
         * http://edndoc.esri.com/arcobjects/9.2/net/shared/geoprocessing/spatial_analyst_tools/how_hillshade_works.htm
         */

        private static final double ALTITUDE_DEG=45f;

        private static final double AZIMUTH_DEG=315f;
        private static final double AZIMUTH_MATH = 360f - AZIMUTH_DEG + 90f;
        private        final double AZIMUTH_RAD = Math.toRadians(AZIMUTH_MATH);

        private static final double ZENITH_DEG=(90d-ALTITUDE_DEG);
        private        final double ZENITH_RAD=Math.toRadians(ZENITH_DEG);
        private        final double ZENITH_COS=Math.cos(ZENITH_RAD);
        private        final double ZENITH_SIN=Math.sin(ZENITH_RAD);


        private final static double DOUBLE_PI=Math.PI*2d;
        private final static double HALF_PI=Math.PI/2d;
        private final static double ONEHALF_PI=DOUBLE_PI-HALF_PI;


        @Override
        public long bgOnProcess(ServiceContext sc) {
            for (int x=0; x<TABLE_DIM; x++) {
                for (int y=0; y<TABLE_DIM; y++) {
                    table[x][y]=hillshade(indexToDelta(x), indexToDelta(y));
                }
            }

            isInitialized = true;
            AppBroadcaster.broadcast(sc.getContext(), AppBroadcaster.FILE_CHANGED_INCACHE, ID);

            return TABLE_SIZE;
        }


        private byte hillshade(float dzx, float dzy) {

            final double slope=slope_rad(dzx, dzy);

            int
                    shade = (int) (255d * (( ZENITH_COS * Math.cos(slope) ) +
                    ( ZENITH_SIN * Math.sin(slope) * Math.cos(AZIMUTH_RAD - aspect_rad(dzx, dzy)) ) ));


            shade = 255-Math.max(MAX_DARKNES, shade);


            return (byte) shade;
        }


        private double slope_rad(final double dzx, final double dzy) {
            return Math.atan(Math.sqrt(dzx*dzx + dzy*dzy));
        }


        private double aspect_rad(final double dzx, final double dzy) {
            double ret=0f;

            if (dzx!=0) {
                ret = Math.atan2(dzy, -1d*dzx);

                if (ret < 0) {

                    ret = DOUBLE_PI + ret;
                }

            } else {
                if (dzy > 0) {
                    ret = HALF_PI;

                } else if (dzy < 0) {
                    ret = ONEHALF_PI;
                }
            }
            return ret;
        }


    }


    public static final Factory FACTORY = new ObjectHandle.Factory() {

        @Override
        public ObjectHandle factory(String id, ServiceContext sc) {
            return  new HillshadeColorTable();
        }
    };
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/elevation/HillShadeColorTable.kt
package ch.bailu.aat_lib.service.cache.elevation

import ch.bailu.aat_lib.lib.color.ARGB
import ch.bailu.aat_lib.service.elevation.tile.MultiCell
import ch.bailu.aat_lib.util.Limit
import kotlin.math.atan
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class HillShadeColorTable {
    private val table = Array(TABLE_DIM) { IntArray(TABLE_DIM) }

    init {
        for (x in table.indices) {
            for (y in table.indices) {
                table[x][y] = hillShadeColor(indexToDelta(x), indexToDelta(y))
            }
        }
    }

    fun getColor(multiCell: MultiCell): Int {
        val x = deltaToIndex(multiCell.deltaZX())
        val y = deltaToIndex(multiCell.deltaZY())
        return table[x][y]
    }

    private fun slopeRad(dzx: Double, dzy: Double): Double {
        return atan(sqrt(dzx * dzx + dzy * dzy))
    }

    private fun aspectRad(dzx: Double, dzy: Double): Double {
        var ret = 0.0

        if (dzx != 0.0) {
            ret = atan2(dzy, -1.0 * dzx)

            if (ret < 0) {
                ret += DOUBLE_PI
            }
        } else {
            if (dzy > 0) {
                ret = HALF_PI
            } else if (dzy < 0) {
                ret = ONEHALF_PI
            }
        }
        return ret
    }

    /**
     * Cairo expects pre-multiplied alpha. Else there will be pixel corruption when
     * saving images.
     */
    private fun preMultipliedAlpha(alpha: Int, color: Int): Int {
        return ((color * alpha).toDouble() / 255.0).toInt()
    }

    private fun hillShadeColor(dzx: Double, dzy: Double): Int {
        val alpha = hillShadeAlpha(dzx, dzy, ALPHA_LIMIT)
        val color = preMultipliedAlpha(alpha, COLOR)
        return ARGB(alpha, color, color, color).toInt()
    }

    private fun hillShadeAlpha(dzx: Double, dzy: Double, limit: Int): Int {
        val slope = slopeRad(dzx, dzy)
        val shade = (limit * ((ZENITH_COS * cos(slope)) +
                (ZENITH_SIN * sin(slope) * cos(AZIMUTH_RAD - aspectRad(dzx, dzy))))).toInt()

        return Limit.clamp(limit - shade, 0, MAX_DARKNESS)
    }

    companion object {
        private const val ALPHA_LIMIT = 175
        private const val DELTA_SCALE = 50.0
        private const val MAX_DARKNESS = 255
        private const val TABLE_DIM = 500
        private const val TABLE_HDIM = TABLE_DIM / 2
        const val TABLE_SIZE = TABLE_DIM * TABLE_DIM

        private const val COLOR = 50
        /**
         * Source:
         * http://edndoc.esri.com/arcobjects/9.2/net/shared/geoprocessing/spatial_analyst_tools/how_hillshade_works.htm
         */
        private const val ALTITUDE_DEG = 45.0

        private const val AZIMUTH_DEG = 315.0
        private const val AZIMUTH_MATH = 360f - AZIMUTH_DEG + 90f
        private const val ZENITH_DEG = (90.0 - ALTITUDE_DEG)
        private const val DOUBLE_PI = Math.PI * 2.0
        private const val HALF_PI = Math.PI * 0.5
        private const val ONEHALF_PI = Math.PI * 1.5

        private val AZIMUTH_RAD = Math.toRadians(AZIMUTH_MATH)
        private val ZENITH_RAD = Math.toRadians(ZENITH_DEG)
        private val ZENITH_COS = cos(ZENITH_RAD)
        private val ZENITH_SIN = sin(ZENITH_RAD)

        private fun indexToDelta(i: Int): Double {
            return (i - TABLE_HDIM).toDouble() / DELTA_SCALE
        }

        private fun deltaToIndex(delta: Int): Int {
            return Limit.clamp(delta, TABLE_HDIM * -1, TABLE_HDIM-1) + TABLE_HDIM
        }
    }
}
